File: AFMobile/AFAndroid/app/src/main/java/cz/cvut/fel/matyapav/afandroid/builders/widgets/OptionWidgetBuilder.java
package cz.cvut.fel.matyapav.afandroid.builders.widgets;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import cz.cvut.fel.matyapav.afandroid.builders.skins.Skin;
import cz.cvut.fel.matyapav.afandroid.components.parts.AFField;
import cz.cvut.fel.matyapav.afandroid.components.parts.FieldInfo;
import cz.cvut.fel.matyapav.afandroid.utils.Localization;

/**
 * @author Pavel Matyáš (matyapav@fel.cvut.cz).
 *
 *@since 1.0.0..
 */
public class OptionWidgetBuilder extends BasicBuilder {

    OptionWidgetBuilder(Context context, Skin skin, FieldInfo properties) {
        super(context, skin, properties);
    }

    @Override
    public View buildFieldView(Context context) {
        RadioGroup radioGroup = new RadioGroup(context);
        radioGroup.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        radioGroup.setOrientation(RadioGroup.VERTICAL); //TODO change according to layout parameters
        int numberOfOptions;
        if(getProperties().getOptions() != null && !getProperties().getOptions().isEmpty()) {
            numberOfOptions = getProperties().getOptions().size();
            RadioButton[] options = new RadioButton[numberOfOptions];
            for (int i = 0; i < numberOfOptions; i++) {
                options[i] = new RadioButton(context);
                options[i].setTextColor(getSkin().getFieldColor());
                options[i].setTypeface(getSkin().getFieldFont());
                options[i].setText(getProperties().getOptions().get(i).getValue());
                options[i].setId(i + 100);
                radioGroup.addView(options[i]);
            }
        }else{
            //YES NO OPTIONS
            numberOfOptions = 2;
            RadioButton[] options = new RadioButton[numberOfOptions];
            options[0] = new RadioButton(context);
            options[0].setTextColor(getSkin().getFieldColor());
            options[0].setTypeface(getSkin().getFieldFont());
            options[0].setText(Localization.translate(getContext(), "option.yes"));
            options[1] = new RadioButton(context);
            options[1].setTextColor(getSkin().getFieldColor());
            options[1].setTypeface(getSkin().getFieldFont());
            options[1].setText(Localization.translate(getContext(), "option.no"));
            radioGroup.addView(options[0]);
            radioGroup.addView(options[1]);
        }
        if(getProperties().isReadOnly()){
            radioGroup.setEnabled(false);
        }
        return radioGroup;
    }

    //TODO REWRITE
    @Override
    public void setData(AFField field, Object value) {

        RadioGroup group = (RadioGroup) field.getFieldView();
        if(value == null){
            group.clearCheck();
            field.setActualData(null);
            return;
        }
        for (int i = 0; i < group.getChildCount(); i++) {
            RadioButton btn = (RadioButton) group.getChildAt(i);
            if(btn.getText().equals(value)
                    || (Boolean.valueOf(value.toString()) && i==0)
                    || (!Boolean.valueOf(value.toString()) && i==1)){
                btn.setChecked(true);
                field.setActualData(value);
                break;
            }
        };

    }

    @Override
    public Object getData(AFField field) {
        RadioGroup group = (RadioGroup) field.getFieldView();
        for (int i = 0; i < group.getChildCount(); i++) {
            RadioButton btn = (RadioButton) group.getChildAt(i);
            if(btn.isChecked()){
                if(btn.getText().toString().equals(Localization.translate(getContext(), "option.yes"))){
                    return true;
                }else if(btn.getText().toString().equals(Localization.translate(getContext(), "option.no"))){
                    return false;
                }else {
                    return btn.getText().toString();
                }
            }
        }
        return null; //nothing is checked
    }
}

File: AFMobile/AFWindowsPhone/AFWindowsPhone/builders/widgets/OptionWidgetBuilder.cs
using AFWindowsPhone.builders.components.parts;
using AFWindowsPhone.builders.skins;
using System;
using Windows.UI.Xaml;
using System.Linq;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;
using AFWindowsPhone.utils;

namespace AFWindowsPhone.builders.widgets
{
    class OptionWidgetBuilder : BasicBuilder
    {
        public OptionWidgetBuilder(Skin skin, FieldInfo properties) : base(skin, properties)
        {
        }

        public override FrameworkElement buildFieldView()
        {
            StackPanel radioGroup = new StackPanel();
            radioGroup.HorizontalAlignment = HorizontalAlignment.Stretch;
            radioGroup.VerticalAlignment = VerticalAlignment.Top;
            radioGroup.Orientation = Orientation.Vertical;
            int numberOfOptions;
            if (getProperties().getOptions() != null && getProperties().getOptions().Any())
            {
                numberOfOptions = getProperties().getOptions().Count;
                RadioButton[] options = new RadioButton[numberOfOptions];
                String groupName = Guid.NewGuid().ToString();
                for (int i = 0; i < numberOfOptions; i++)
                {
                    options[i] = new RadioButton();
                    options[i].FontFamily = getSkin().getFieldFont();
                    options[i].FontSize = getSkin().getFieldFontSize();
                    options[i].Content = getProperties().getOptions()[i].getValue();
                    options[i].GroupName = groupName;
                    if (getProperties().isReadOnly())
                    {
                        options[i].IsEnabled = false;
                    }
                    radioGroup.Children.Add(options[i]);
                }
            }
            else {
                //YES NO OPTIONS
                numberOfOptions = 2;
                RadioButton[] options = new RadioButton[numberOfOptions];
                options[0] = new RadioButton();
                options[0].FontFamily = getSkin().getFieldFont();
                options[0].FontSize = getSkin().getFieldFontSize();
                options[0].Content = Localization.translate("option.yes");
                options[1] = new RadioButton();
                options[1].FontFamily = getSkin().getFieldFont();
                options[1].FontSize = getSkin().getFieldFontSize();
                options[1].Content = Localization.translate("option.no");
                if (getProperties().isReadOnly())
                {
                    options[0].IsEnabled = false;
                    options[1].IsEnabled = false;
                }
                radioGroup.Children.Add(options[0]);
                radioGroup.Children.Add(options[1]);
            }
          
            return radioGroup;
        }

        public override void setData(AFField field, Object value)
        {

            StackPanel group = (StackPanel) field.getFieldView();
            if (value == null)
            {
                foreach (RadioButton btn in group.Children) //TODO asi hodi chybu
                {
                    btn.IsChecked = false;
                }
                field.setActualData(value);
                return;
            }
            for (int i = 0; i < group.Children.Count; i++)
            {
                RadioButton btn = (RadioButton) group.Children[i];
                if (btn.Content.Equals(value)
                        || (Utils.TryToConvertIntoBoolean(value.ToString()) == true && i == 0)
                        || (Utils.TryToConvertIntoBoolean(value.ToString()) == false && i == 1))
                {
                    btn.IsChecked = true;
                    field.setActualData(value);
                    break;
                }
            };

        }

        public override Object getData(AFField field)
        {
            StackPanel group = (StackPanel) field.getFieldView();
      
            for (int i = 0; i < group.Children.Count; i++)
            {
                RadioButton btn = (RadioButton) group.Children[i];
                if (btn.IsChecked.Value)
                {
                    if (btn.Content.Equals(Localization.translate("option.yes")))
                    {
                        return "true";
                    }
                    else if (btn.Content.Equals(Localization.translate("option.no")))
                    {
                        return "false";
                    }
                    else {
                        return btn.Content;
                    }
                }
            }
            return null; //nothing is checked
        }
    }
}
